src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs

```diff
@@ -0,0 +1,163 @@
+            var userInfoEndpoint = string.Format(CultureInfo.InvariantCulture, Options.UserInformationEndpoint, contextId);
+
+            var idToken = tokens.Response.RootElement.GetString("id_token");
+
+            if (Options.ValidateTokens)
+            {
+                var validateIdContext = new SuperOfficeValidateIdTokenContext(Context, Scheme, Options, idToken);
+                await Options.Events.ValidateIdToken(validateIdContext);
+            }
+
+            if (Options.SaveTokens)
+            {
+                // Save id_token as well.
+                SaveIdToken(properties, idToken);
+            }
+
+            if (Options.IncludeIdTokenAsClaims)
+            {
+                var tokenClaims = ExtractClaimsFromToken(idToken);
+
+                foreach (var claim in tokenClaims)
+                {
+                    // may be possible same claim names from UserInformationEndpoint and IdToken.
+                    if (!identity.HasClaim(c => c.Type == claim.Type))
+                    {
+                        identity.AddClaim(claim);
+                    }
+                }
+            }
+
+            // Get the SuperOffice user principal
+            using var request = new HttpRequestMessage(HttpMethod.Get, userInfoEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException($"An error occurred when retrieving SuperOffice user information ({response.StatusCode}). Please check if the authentication information is correct and the corresponding SuperOffice API is enabled.");
+            }
+
+            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+            var context = new OAuthCreatingTicketContext(new ClaimsPrincipal(identity), properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+
+            if (Options.IncludeFunctionalRightsAsClaims)
+            {
+                Options.ClaimActions.MapJsonKey(SuperOfficeAuthenticationConstants.PrincipalNames.FunctionRights, SuperOfficeAuthenticationConstants.PrincipalNames.FunctionRights);
```

Comment:
This would seem to add a new duplicate claim action every time a user logs in.

Reply:
Can you point me to where this validation is happening? We shouldn't be mutating the options based on the current user. If you need to conditionally add a claim based on the current user, you might need to add a custom claim action like [`AppleEmailClaimAction`](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.Apple/AppleEmailClaimAction.cs) does.

Reply:
No, there are no special per-user policy. This should be for all users.

Reply:
I have updated the project with a ClaimAction, and tests.